### PR TITLE
6587699: Document DigestInputStream behavior when skip() or mark() / reset() is used

### PR DESCRIPTION
**Pull request by wangweij**
Add an `@implNote` to clarify the behavior when these methods are called. A new test is added.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-6587699](https://bugs.openjdk.org/browse/JDK-6587699): Document DigestInputStream behavior when skip() or mark() / reset() is used
 * [JDK-8291516](https://bugs.openjdk.org/browse/JDK-8291516): Document DigestInputStream behavior when skip() or mark() / reset() is used (**CSR**)

### Reviewers
 * [Bradford Wetmore](https://openjdk.org/census#wetmore) (@bradfordwetmore - **Reviewer**) ⚠️ Review applies to [a6304352](https://git.openjdk.org/jdk/pull/9667/files/a6304352b14418cf75aa0001c3faf3ed95d745df)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9667/head:pull/9667` \
`$ git checkout pull/9667`

Update a local copy of the PR: \
`$ git checkout pull/9667` \
`$ git pull https://git.openjdk.org/jdk pull/9667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9667`

View PR using the GUI difftool: \
`$ git pr show -t 9667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9667.diff">https://git.openjdk.org/jdk/pull/9667.diff</a>

</details>

### CODE REVIEW

**Comment by bradfordwetmore** [comment]
Be sure to update the commit message.  :)  

Hope to look at the rest later.

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/java/security/DigestInputStream.java:55]: "when it is turned on"?  You could probably just end with "actually read from the input stream."

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/java/security/DigestInputStream.java:56]: includes the various

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/java/security/DigestInputStream.java:57]: Specifically
->
Additionally ?

File: test/jdk/java/security/DigestInputStream/TestSkipAndReset.java
Comment (bradfordwetmore) [inline test/jdk/java/security/DigestInputStream/TestSkipAndReset.java:62]: Very clear test, thank you.

**Review by bradfordwetmore — COMMENTED** [review]
I'll approve this after CSR is approved.

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [reply src/java.base/share/classes/java/security/DigestInputStream.java:54]: My feeling still remains.  I don't recall seeing a wording construction like this except for IFF, which is well-understood.  I think it could be said more compactly with:

`This implementation only updates the message digest with data actually read from the input stream when it is on. `

Just my $0.02.

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [reply src/java.base/share/classes/java/security/DigestInputStream.java]: I meant additionally as in "here's an additional case to be aware of...", not "this is an additional read case..."

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/java/security/DigestInputStream.java:57]: Maybe just "Note" instead of "Especially" (or just remove) as this doesn't seem to be an extra special situation.

**Review by bradfordwetmore — APPROVED** [review]
Take or leave.  

Otherwise LGTM.

File: src/java.base/share/classes/java/security/DigestInputStream.java
Comment (jddarcy) [reply src/java.base/share/classes/java/security/DigestInputStream.java:54]: For your consideration, something like:
 "... This implementation will update the message digest if and only if the stream is on ...."

**Comment by bradfordwetmore** [comment]
I don't think a new release is needed.

**Pull request closed without merging**
